pkgs/servers/mastodon/default.nix

```
@@ -0,0 +1,109 @@
+{ nodejs-slim, mkYarnPackage, fetchFromGitHub, bundlerEnv,
+  stdenv, yarn, callPackage, imagemagick, ffmpeg, file,
+
+  # Allow building a fork or custom version of Mastodon:
+  pname ? "mastodon",
+  version ? import ./version.nix,
+  srcOverride ? null,
+  dependenciesDir ? ./.,  # Should contain gemset.nix, yarn.nix and package.json.
+
+  ... }:
+
+let
+in stdenv.mkDerivation rec {
+  inherit pname version;
+
+  # Using overrideAttrs on src does not build the gems and modules with the overridden src.
+  # Putting the callPackage up in the arguments list also does not work.
+  src = if srcOverride != null then srcOverride else callPackage ./source.nix {};
+
+  mastodon-gems = bundlerEnv {
+    name = "${pname}-gems-${version}";
+    inherit version;
+    gemdir = src;
+    gemset = dependenciesDir + "/gemset.nix";
+    # This fix (copied from https://github.com/NixOS/nixpkgs/pull/76765) replaces the gem
+    # symlinks with directories, resolving this error when running rake:
+    #   /nix/store/451rhxkggw53h7253izpbq55nrhs7iv0-mastodon-gems-3.0.1/lib/ruby/gems/2.6.0/gems/bundler-1.17.3/lib/bundler/settings.rb:6:in `<module:Bundler>': uninitialized constant Bundler::Settings (NameError)
+    postBuild = ''
+      for gem in "$out"/lib/ruby/gems/*/gems/*; do
+        cp -a "$gem/" "$gem.new"
+        rm "$gem"
+        # needed on macOS, otherwise the mv yields permission denied
+        chmod +w "$gem.new"
+        mv "$gem.new" "$gem"
+      done
+    '';
+  };
+
+  mastodon-js-modules = mkYarnPackage {
+    pname = "${pname}-modules";
+    yarnNix = dependenciesDir + "/yarn.nix";
+    packageJSON = dependenciesDir + "/package.json";
+    inherit src version;
+  };
+
+  mastodon-assets = stdenv.mkDerivation {
+    pname = "${pname}-assets";
+    inherit src version;
+
+    buildInputs = [
+      mastodon-gems nodejs-slim yarn
+    ];
+
+    buildPhase = ''
+      # Support Mastodon forks which don't call themselves 'mastodon' or which
+      # omit the organization name from package.json.
+      if [ "$(ls ${mastodon-js-modules}/libexec/* | grep node_modules)" ]; then
+          cp -r ${mastodon-js-modules}/libexec/*/node_modules node_modules
+      else
+          cp -r ${mastodon-js-modules}/libexec/*/*/node_modules node_modules
+      fi
+      chmod -R u+w node_modules
+      rake assets:precompile
```

```suggestion
      rake webpacker:compile
```

doesn't try to run `yarn install`

---

Thanks for the suggestion. I had to make the same change in the test which runs Mastodon's test suite.